### DISCUSSION TITLE
Other HTTP response status codes

### DISCUSSION THREAD

**Discussion by thepeterkovacs** [trpc/trpc#4833] [Q&A]
Is there a way to respond with a custom HTTP status code (201, 204, 418...) using a TRPC procedure?
I noticed, that by default a success always returns 200 and the **TRPCError** class only covers a small variety of codes.

I guess the error status code is customizable in the **errorFormatter**, but I feel like there should be a better approach.

**Comment by KATT** [comment]
You should be able to extend TRPCError and do your own mapper for the error formatter if you want more codes

**Reply by thepeterkovacs** [reply]
Thanks! Do you have any example code, or repository that implements this properly and which I can use as a reference?
Also this still doesn't solve the whole problem, since 20X codes are not exceptions, so it is not ideal to pass them through by throwing a TRPCError.

**Reply by KATT** [reply]
I really don't understand why you'd need **all** of them, adding another 2xx would be needed to done by the lib so we'd need strong motivation before even considering that.

You can see a custom error code set here: https://github.com/trpc/trpc/blob/2212b7cd6f304938c9142fc7e7799a13cda19035/packages/tests/server/errors.test.ts#L236-L269

**Reply by thepeterkovacs** [reply]
Not really a strong motivation, but if someone wanted to rearchitect a highly complicated API to use tRPC, that already had all the successful response codes correctly (201 for mutation, 206 for partial data...), it can be a pain to refactor all of that in clients, tests and so on to just be plain 200 for everything.

Not sure how hard it is to implement into the library, but seems like a nice feature to be customizable, since it is the standard way of doing things.

For the error, as I mentioned, it's only a workaround reassigning the status using the **errorFormatter**.
Do you have an example, where the **TRPCError** class is being extended and used correctly to be able to throw other status codes from inside a procedure?

Thanks!

**Reply by siawyoung** [reply]
I would also like this feature.

**Reply by mwimbledon** [reply]
Would be great to see specific success codes implemented in tRPC

**Comment by klesun** [comment]
This seems to work for me:
```ts
const t = initTRPC.context<Context>().create({
  errorFormatter({ shape, error }) {
    shape.data.httpStatus = 503;
    return shape;
  }
}
```
<img width="346" height="89" alt="image" src="https://github.com/user-attachments/assets/a6bf1d23-355d-4dc4-824d-b3a27de7888c" />
